### PR TITLE
Auth: Lock down Grafana admin role updates if the role is externally synced

### PR DESCRIPTION
**Pull request by IevaVasiljeva**
**What is this feature?**

Roles should either be manually updated or externally synced. If they are externally synced, we lock manual role updates.

**Why do we need this feature?**

For clearer and more secure role management.

**Who is this feature for?**

Anyone who uses external auth providers to sync user's roles.

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/api/user.go [hunk 1]
```diff
@@ -68,6 +68,11 @@ func (hs *HTTPServer) getUserUserProfile(c *contextmodel.ReqContext, userID int6
 		userProfile.AuthLabels = append(userProfile.AuthLabels, authLabel)
 		userProfile.IsExternal = true
 		userProfile.IsExternallySynced = login.IsExternallySynced(hs.Cfg, authInfo.AuthModule)
+		oAuthExternallySynced := false
+		if oauthInfo := hs.SocialService.GetOAuthInfoProvider(strings.TrimPrefix(authInfo.AuthModule, "oauth_")); oauthInfo != nil {
+			oAuthExternallySynced = oauthInfo.AllowAssignGrafanaAdmin
+		}
```
Comment (IevaVasiljeva) [inline pkg/api/user.go:74]: note: annoyingly, I couldn't pass in the whole `oauthInfo` or `SocialService`, as that creates a circular dependency.

File: pkg/services/login/authinfo.go [hunk 2]
```diff
@@ -114,6 +114,19 @@ func IsExternallySynced(cfg *setting.Cfg, authModule string) bool {
...
+	if !IsExternallySynced(cfg, authModule) {
+		return false
+	}
+
+	switch authModule {
+	case JWTModule:
+		return cfg.JWTAuthAllowAssignGrafanaAdmin
+	default:
+		return oAuthAndAllowAssignGrafanaAdmin
+	}
```
Comment (IevaVasiljeva) [inline pkg/services/login/authinfo.go:127]: question: am I correct in assuming that Grafana Admin role can only be synced for JWT and OAuth providers?

I'm also not keen on the variable naming here, but couldn't come up with anything better.

File: pkg/api/admin_users.go [hunk 3]
```diff
@@ -167,6 +167,19 @@ func (hs *HTTPServer) AdminUpdateUserPermissions(c *contextmodel.ReqContext) res
 		return response.Error(http.StatusBadRequest, "id is invalid", err)
 	}
 
+	getAuthQuery := login.GetAuthInfoQuery{UserId: userID}
+	if authInfo, err := hs.authInfoService.GetAuthInfo(c.Req.Context(), &getAuthQuery); err == nil {
+		oAuthAndAllowAssignGrafanaAdmin := false
+		if oauthInfo := hs.SocialService.GetOAuthInfoProvider(strings.TrimPrefix(authInfo.AuthModule, "oauth_")); oauthInfo != nil {
+			oAuthAndAllowAssignGrafanaAdmin = oauthInfo.AllowAssignGrafanaAdmin
+		}
```
Comment (IevaVasiljeva) [inline pkg/api/admin_users.go:175]: note: annoyingly, I couldn't pass in the whole oauthInfo or SocialService, as that creates a circular dependency.

File: pkg/services/login/authinfo.go (hunk 2 above)
Comment (Jguer) [reply pkg/services/login/authinfo.go:127]: I think only JWT, OAuth and LDAP

Comment (IevaVasiljeva) [reply pkg/services/login/authinfo.go:132]: Thanks for the heads up about LDAP, I had missed that one.

**Review by Jguer — APPROVED** [review]

**Pull request merged**